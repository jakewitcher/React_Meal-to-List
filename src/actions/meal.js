import uuid from 'uuid';

export const addMeal = (
    {
        name = '',
        itemList = [],
    } = {}
) => ({
    type: 'ADD_MEAL',
    meal: {
        id: uuid(),
        name,
        itemList,
    }
});

export const deleteMeal = (id = '') => ({
    type: 'DELETE_MEAL',
    id,
});

export const editMeal = (
    {
        name = '',
        id = '',
        itemList = [],
    } = {}
) => ({
    type: 'EDIT_MEAL',
    meal: {
        id,
        name,
        itemList,
    }
});