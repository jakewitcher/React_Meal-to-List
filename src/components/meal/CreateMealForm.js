import React from 'react';

const MealForm = (props) => {
    const {
        mealNameChange,
        itemNameChange,
        amountChange,
        unitChange,
        addItem,
        resetMeal,
        mealName,
        itemName,
        amount,
        unit,
        itemList,
        dispatch
    } = props.mealProps;
    const { addMeal } = props;

    return (
        <div className="form">
            <h2 className="form__header">Create a New Meal</h2>
            <div className="form__form">
                <div className="form__nameinputfield">
                    <p>Meal Name</p>
                    <input
                        className="form__nameinputfield--input"
                        type="text"
                        value={mealName}
                        onChange={mealNameChange}
                    />
                </div>
                <div className="form__iteminputfields">
                    <div>
                        <p>Item</p>
                        <input
                            className="form__iteminputfields--input"
                            type="text"
                            value={itemName}
                            onChange={itemNameChange}
                        />
                    </div>

                    <div>
                        <p>Amount</p>
                        <input
                            className="form__iteminputfields--input"
                            type="number"
                            value={amount}
                            onChange={amountChange}
                        />
                    </div>

                    <div>
                        <p>Unit</p>
                        <select
                            className="form__iteminputfields--input form__iteminputfields--dropdown"
                            name="unit"
                            value={unit}
                            onChange={unitChange}
                        >
                            <option value="pound(s)">pound(s)</option>
                            <option value="ounce(s)">ounce(s)</option>
                            <option value="gallon(s)">gallons</option>
                            <option value="quart(s)">quart(s)</option>
                            <option value="pint(s)">pint(s)</option>
                            <option value="item(s)">item(s)</option>
                            <option value="bag(s)">bag(s)</option>
                            <option value="box(es)">box(es)</option>
                            <option value="container(s)">container(s)</option>
                        </select>
                    </div>
                    <div className="form__button">
                        <button
                            className="form__button--add"
                            onClick={addItem}
                        >
                            Add Item
                        </button>
                    </div>
                </div>
                <div>
                    <button
                        className="form__button--save"
                        onClick={() => {
                            const name = mealName;
                            dispatch(addMeal({
                                name,
                                itemList
                            }));
                            resetMeal();
                        }}
                    >
                        Save Meal
                    </button>
                </div>
            </div>
        </div>
    )
}



export default MealForm;