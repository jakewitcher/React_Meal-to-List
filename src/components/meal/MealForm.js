import React from 'react';

const MealForm = (props) => {
    const {
        mealNameChange,
        itemNameChange,
        amountChange,
        unitChange,
        addItem,
        resetMeal,
        mealName,
        itemName,
        amount,
        unit,
        itemList,
        updateMeal,
        mealId,
        title,
    } = props;

    const pluralUnitValue = (amount, unitType) => {
        if (amount < 2) {
            return `${unitType}`
        }
        return `${unitType}s`
    };

    return (
        <div className="form">
            <h2 className="form__header">{title}</h2>
            <div className="form__form">
                <div className="form__nameinputfield">
                    <p>Meal Name</p>
                    <input
                        className="form__nameinputfield--input"
                        type="text"
                        value={mealName}
                        onChange={mealNameChange}
                    />
                </div>
                <div className="form__iteminputfields">
                    <div>
                        <p>Item</p>
                        <input
                            className="form__iteminputfields--input"
                            type="text"
                            value={itemName}
                            onChange={itemNameChange}
                        />
                    </div>

                    <div>
                        <p>Amount</p>
                        <input
                            className="form__iteminputfields--input"
                            type="number"
                            value={amount}
                            onChange={amountChange}
                        />
                    </div>

                    <div>
                        <p>Unit</p>
                        <select
                            className="form__iteminputfields--input form__iteminputfields--dropdown"
                            name="unit"
                            value={unit}
                            onChange={unitChange}
                        >
                            <option value={pluralUnitValue(amount, 'pound')}>{pluralUnitValue(amount, 'pound')}</option>
                            <option value={pluralUnitValue(amount, 'ounce')}>{pluralUnitValue(amount, 'ounce')}</option>
                            <option value={pluralUnitValue(amount, 'gallon')}>{pluralUnitValue(amount, 'gallon')}</option>
                            <option value={pluralUnitValue(amount, 'quart')}>{pluralUnitValue(amount, 'quart')}</option>
                            <option value={pluralUnitValue(amount, 'pint')}>{pluralUnitValue(amount, 'pint')}</option>
                            <option value={pluralUnitValue(amount, 'item')}>{pluralUnitValue(amount, 'item')}</option>
                            <option value={pluralUnitValue(amount, 'bag')}>{pluralUnitValue(amount, 'bag')}</option>
                            <option value={pluralUnitValue(amount, 'container')}>{pluralUnitValue(amount, 'container')}</option>
                        </select>
                    </div>
                    <div className="form__button">
                        <button
                            className="form__button--add"
                            onClick={addItem}
                        >
                            Add Item
                        </button>
                    </div>
                </div>
                <div>
                    <button
                        className="form__button--save"
                        onClick={() => {
                            const name = mealName;
                            const id = mealId;
                            updateMeal({
                                name,
                                itemList,
                                id,
                            });
                            resetMeal();
                        }}
                    >
                        Save Meal
                    </button>
                </div>
            </div>
        </div>
    )
}

export default MealForm;