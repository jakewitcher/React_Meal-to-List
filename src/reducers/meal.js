import { defaultMealsList } from './seed';

// const mealDefaultState = {
//     mealList: [],
// };

export default (state = {mealList: defaultMealsList}, action) => {
    switch (action.type) {
        case 'ADD_MEAL':
            return {
                ...state,
                mealList: [...state.mealList, action.meal]
            };
        default:
            return state;
    }
}