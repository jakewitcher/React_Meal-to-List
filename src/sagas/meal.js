import { put, takeEvery, select } from 'redux-saga/effects';
import database from '../firebase/firebase';

const getUserId = (state) => state.auth.uid;

const formatItemList = (list) => {
    return list.reduce((list, item) => {
        list[item.itemName] = {
            amount: item.amount,
            unit: item.unit
        };
        return list;
    }, {});
};

function* addMeal({ meal = {} } = {}) {
    const { itemList, name } = meal;
    const items = formatItemList(itemList);

    let id;
    const uid = yield select(getUserId);
    yield database.ref(`users/${uid}/meals`).push({ name, items }).then((ref) => {
        id = ref.key;
    });

    yield put({
        type: 'ADD_MEAL',
        meal: {
            name,
            itemList,
            id,
        }
    });
}

function* editMeal({ meal = {} }) {
    const { itemList, name, id } = meal;
    const items = formatItemList(itemList);
    const uid = yield select(getUserId);

    yield database.ref(`users/${uid}/meals/${id}`).update({ name, items });

    yield put({
        type: 'EDIT_MEAL',
        meal: {
            name,
            itemList,
            id,
        }
    });
};

function* deleteMeal({ id = '' }) {
    const uid = yield select(getUserId);

    yield database.ref(`users/${uid}/meals/${id}`).remove();

    yield put({
        type: 'DELETE_MEAL',
        id,
    });
}

function* setMeal() {
    const meals = [];
    const uid = yield select(getUserId);

    yield database.ref(`users/${uid}/meals`).once('value').then((snapshot) => {
        snapshot.forEach((childSnapshot) => {
            const itemList = [];
            childSnapshot.child('items').forEach((item) => {
                itemList.push({
                    itemName: item.key,
                    amount: item.val().amount,
                    unit: item.val().unit,
                });
            });

            meals.push({
                id: childSnapshot.key,
                name: childSnapshot.val().name,
                itemList,
            });
        });
    });

    yield put({ type: 'SET_MEAL', meals });
};

export function* watchAddMeal() {
    yield takeEvery('ON_ADD_MEAL', addMeal);
};

export function* watchEditMeal() {
    yield takeEvery('ON_EDIT_MEAL', editMeal);
};

export function* watchDeleteMeal() {
    yield takeEvery('ON_DELETE_MEAL', deleteMeal);
};

export function* watchSetMeal() {
    yield takeEvery('ON_SET_MEAL', setMeal);
}