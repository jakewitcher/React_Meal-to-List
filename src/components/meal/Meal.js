import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { onDeleteMeal } from '../../actions/meal';

export const Meal = (props) => {
    const { name, id, itemList } = props;
    return (
        <div className="list-type">
            <Link className="list-type__link" to={`/meals/meal/${id}`}>
                <p className="list-type__name">{name}</p>
            </Link>
            <p className="list-type__details">number of items: <span className="list-type__details list-type__details--span"> {itemList.length}</span></p>
            <div className="list-type__update">
                <Link className="list-type__link" to={`/meals/edit/${id}`}>
                    <i className="list-type__edit fas fa-edit" />
                </Link>
                <i
                    className="list-type__delete fas fa-trash"
                    onClick={() => props.onDeleteMeal(id)}
                />
            </div>
        </div>
    );
}

const mapDispatchToProps = (dispatch) => {
    return {
        onDeleteMeal: (id) => dispatch(onDeleteMeal(id)),
    }
}

export default connect(undefined, mapDispatchToProps)(Meal);