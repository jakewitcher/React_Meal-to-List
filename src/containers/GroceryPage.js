import React, { Component } from 'react';
import { connect } from 'react-redux';
import GroceryForm from '../components/grocery/GroceryForm';
import GroceryItems from '../components/grocery/GroceryItems';
import { addGrocery, editGrocery } from '../actions/grocery';

class GroceryPage extends Component {
    constructor(props) {
        super(props);
        this.state = {
            groceryName: props.grocery ? props.grocery.name : '',
            selectedMeal: this.props.meals.mealList[0] || {},
            itemsMap: props.grocery ? groceryMap(props.grocery) : new Map(),
            itemsList: props.grocery ? props.grocery.items : [],
        }
    }

    deleteGroceryItem = (id) => {
        const newGroceryList = this.state.itemsList.filter(item => item.id !== id);
        const newGroceryMap = new Map();
        newGroceryList.forEach((item) => {
            newGroceryMap.set(item.itemName, item);
        });
        this.setState({
            itemsList: newGroceryList,
            itemsMap: newGroceryMap,
        });
    };

    addMealToList = () => {
        const { itemsMap, selectedMeal } = this.state;
        const CombinedItemsList = selectedMeal.itemList.reduce((a, b) => {
            if (a.has(b.itemName)) {
                const newAmount = Number(a.get(b.itemName).amount) + Number(b.amount);
                const itemCopy = Object.assign({}, b)
                itemCopy.amount = newAmount;
                a.set(itemCopy.itemName, itemCopy);
                return a;
            }
            a.set(b.itemName, b);
            return a;
        }, itemsMap);

        const newItemsList = [];
        CombinedItemsList.forEach((i => newItemsList.push(i)));

        this.setState({
            itemsMap: CombinedItemsList,
            itemsList: newItemsList,
            selectedMeal: this.props.meals.mealList[0],
        });
    };

    selectedMealChange = (e) => {
        const { meals } = this.props;
        const selectedMealName = e.target.value;
        const selectedMeal = meals.mealList.filter(meal => meal.name === selectedMealName);
        this.setState({
            selectedMeal: selectedMeal[0],
        });
    };

    groceryNameChange = (e) => {
        const groceryName = e.target.value;
        this.setState({ groceryName });
    };

    resetGrocery = () => {
        this.setState({ groceryName: "", itemsList: [], itemsMap: new Map() });
    }
    render() {
        return (
            <div>
                <div className="grocery">
                    <GroceryForm
                        selectedMeal={this.state.selectedMeal}
                        groceryName={this.state.groceryName}
                        itemsList={this.state.itemsList}
                        selectedMealChange={this.selectedMealChange}
                        groceryNameChange={this.groceryNameChange}
                        addMealToList={this.addMealToList}
                        resetGrocery={this.resetGrocery}
                        meals={this.props.meals}
                        groceryId={this.props.grocery ? this.props.grocery.id : ''}
                        title={this.props.grocery ? 'Edit Grocery List' : 'Create Grocery List'}
                        updateGrocery={this.props.grocery ? this.props.editGrocery : this.props.addGrocery}
                    />
                    <GroceryItems
                        groceryName={this.state.groceryName}
                        itemsList={this.state.itemsList}
                        deleteGroceryItem={this.deleteGroceryItem}
                    />
                </div>
            </div>

        )
    }

};

const groceryMap = (grocery) => {
    const selectedGroceryMap = new Map();
    grocery.items.forEach((item) => {
        selectedGroceryMap.set(item.itemName, item);
    });
    return selectedGroceryMap;
};

const mapStateToProps = (state, props) => {
    return {
        groceryLists: state.groceryLists,
        meals: state.meals,
        grocery: state.groceryLists.groceryList.filter((list) => list.id === props.match.params.groceryId)[0],
    }
}

const mapDispatchToProps = (dispatch) => {
    return {
        editGrocery: (data) => dispatch(editGrocery(data)),
        addGrocery: (data) => dispatch(addGrocery(data)),
    }
}

export default connect(mapStateToProps, mapDispatchToProps)(GroceryPage);