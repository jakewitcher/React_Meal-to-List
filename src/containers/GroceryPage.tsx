import React, { Component, ChangeEvent } from "react";
import { Dispatch } from "redux";
import { connect } from "react-redux";
import GroceryForm from "../components/GroceryForm";
import GroceryItems from "../components/GroceryItems";
import { onAddGrocery, onEditGrocery } from "../store/grocery/actions";
import { Grocery, GroceryMeals } from "../store/grocery/types";
import { MealState, Meal } from "../store/meal/types";
import { Item } from "../store/item/types";
import { ApplicationState } from "../store";
import { RouteComponentProps } from "react-router";

interface IGroceryPageProps extends RouteComponentProps<{ groceryId: string }> {
  grocery?: Grocery;
  meals: MealState;
  onAddGrocery(grocery: Grocery): void;
  onEditGrocery(grocery: Grocery): void;
}

interface IGroceryPageState {
  groceryName: string;
  selectedMeal: Meal;
  itemList: Item[];
  groceryListMeals: GroceryMeals;
}

export class GroceryPage extends Component<
  IGroceryPageProps,
  IGroceryPageState
> {
  constructor(props: IGroceryPageProps) {
    super(props);
    this.state = {
      groceryName: props.grocery ? props.grocery.name : "",
      selectedMeal: props.meals.mealList[0] || {},
      itemList: props.grocery ? props.grocery.itemList : [],
      groceryListMeals: props.grocery ? props.grocery.groceryListMeals : {}
    };
  }

  deleteGroceryItem = (name: string) => {
    const newGroceryList = this.state.itemList.filter(
      item => item.itemName !== name
    );
    this.setState({
      itemList: newGroceryList
    });
  };

  addMealToList = () => {
    const { itemList, selectedMeal, groceryListMeals } = this.state;
    const listCopy = itemList.map(i => Object.assign({}, i));

    const combinedItemList = selectedMeal.itemList.reduce((list, item) => {
      const itemFound = list.find(i => i.itemName === item.itemName);
      if (itemFound) {
        itemFound.amount = item.amount + itemFound.amount;
        return list;
      }

      list.push(item);
      return list;
    }, listCopy);

    const updatedGroceryListMeals = {
      ...groceryListMeals,
      [selectedMeal.id]: (groceryListMeals[selectedMeal.id] || 0) + 1
    };

    this.setState({
      itemList: combinedItemList,
      groceryListMeals: updatedGroceryListMeals
    });
  };

  selectedMealChange = (e: ChangeEvent<HTMLSelectElement>) => {
    const { meals } = this.props;
    const selectedMealName = e.target.value;
    const selectedMeal = meals.mealList.filter(
      meal => meal.name === selectedMealName
    );
    this.setState({
      selectedMeal: selectedMeal[0]
    });
  };

  groceryNameChange = (e: ChangeEvent<HTMLInputElement>) => {
    const groceryName = e.target.value;
    this.setState({ groceryName });
  };

  resetGrocery = () => {
    this.setState({ groceryName: "", itemList: [], groceryListMeals: {} });
  };
  render() {
    return (
      <div>
        <div className="update--grocery">
          <GroceryForm
            selectedMeal={this.state.selectedMeal}
            groceryName={this.state.groceryName}
            itemList={this.state.itemList}
            groceryListMeals={this.state.groceryListMeals}
            selectedMealChange={this.selectedMealChange}
            groceryNameChange={this.groceryNameChange}
            addMealToList={this.addMealToList}
            resetGrocery={this.resetGrocery}
            meals={this.props.meals}
            groceryId={this.props.grocery ? this.props.grocery.id : ""}
            updateGrocery={
              this.props.grocery
                ? this.props.onEditGrocery
                : this.props.onAddGrocery
            }
          />
          <GroceryItems
            groceryName={this.state.groceryName}
            itemList={this.state.itemList}
            deleteGroceryItem={this.deleteGroceryItem}
            isDetails={false}
          />
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state: ApplicationState, props: IGroceryPageProps) => {
  return {
    groceryLists: state.groceryLists,
    meals: state.meals,
    grocery: state.groceryLists.groceryList.filter(
      list => list.id === props.match.params.groceryId
    )[0]
  };
};

const mapDispatchToProps = (dispatch: Dispatch) => {
  return {
    onEditGrocery: (data: Grocery) => dispatch(onEditGrocery(data)),
    onAddGrocery: (data: Grocery) => dispatch(onAddGrocery(data))
  };
};

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(GroceryPage);
