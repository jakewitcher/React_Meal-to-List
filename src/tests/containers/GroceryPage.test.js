import React from 'react';
import { shallow } from 'enzyme';
import { GroceryPage } from '../../containers/GroceryPage';
import { mealsList } from '../fixtures/meal';

test('should render MealPage correctly', () => {
    const meals = { mealList: mealsList };
    const wrapper = shallow(<GroceryPage meals={meals} />);
    expect(wrapper).toMatchSnapshot();
});