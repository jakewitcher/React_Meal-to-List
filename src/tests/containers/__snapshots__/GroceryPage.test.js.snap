// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render MealPage correctly 1`] = `
<div>
  <div
    className="grocery"
  >
    <GroceryForm
      addMealToList={[Function]}
      groceryId=""
      groceryName=""
      groceryNameChange={[Function]}
      itemList={Array []}
      meals={
        Object {
          "mealList": Array [
            Object {
              "id": 101,
              "itemList": Array [
                Object {
                  "amount": 2,
                  "id": 1,
                  "itemName": "bacon",
                  "unit": "pounds",
                },
                Object {
                  "amount": 12,
                  "id": 2,
                  "itemName": "eggs",
                  "unit": "items",
                },
                Object {
                  "amount": 1,
                  "id": 3,
                  "itemName": "toast",
                  "unit": "bag",
                },
                Object {
                  "amount": 1,
                  "id": 4,
                  "itemName": "orange juice",
                  "unit": "quart",
                },
              ],
              "name": "breakfast",
            },
            Object {
              "id": 102,
              "itemList": Array [
                Object {
                  "amount": 1,
                  "id": 5,
                  "itemName": "salad",
                  "unit": "bag",
                },
                Object {
                  "amount": 2,
                  "id": 6,
                  "itemName": "chicken",
                  "unit": "pounds",
                },
                Object {
                  "amount": 12,
                  "id": 7,
                  "itemName": "walnuts",
                  "unit": "ounces",
                },
                Object {
                  "amount": 4,
                  "id": 8,
                  "itemName": "apples",
                  "unit": "items",
                },
                Object {
                  "amount": 1,
                  "id": 1,
                  "itemName": "bacon",
                  "unit": "pound",
                },
              ],
              "name": "lunch",
            },
            Object {
              "id": 103,
              "itemList": Array [
                Object {
                  "amount": 1,
                  "id": 9,
                  "itemName": "pasta",
                  "unit": "container",
                },
                Object {
                  "amount": 2,
                  "id": 10,
                  "itemName": "marinara sauce",
                  "unit": "pints",
                },
                Object {
                  "amount": 12,
                  "id": 11,
                  "itemName": "breadsticks",
                  "unit": "items",
                },
                Object {
                  "amount": 1,
                  "id": 5,
                  "itemName": "salad",
                  "unit": "bag",
                },
              ],
              "name": "dinner",
            },
          ],
        }
      }
      resetGrocery={[Function]}
      selectedMeal={
        Object {
          "id": 101,
          "itemList": Array [
            Object {
              "amount": 2,
              "id": 1,
              "itemName": "bacon",
              "unit": "pounds",
            },
            Object {
              "amount": 12,
              "id": 2,
              "itemName": "eggs",
              "unit": "items",
            },
            Object {
              "amount": 1,
              "id": 3,
              "itemName": "toast",
              "unit": "bag",
            },
            Object {
              "amount": 1,
              "id": 4,
              "itemName": "orange juice",
              "unit": "quart",
            },
          ],
          "name": "breakfast",
        }
      }
      selectedMealChange={[Function]}
      title="Create Grocery List"
    />
    <GroceryItems
      deleteGroceryItem={[Function]}
      groceryName=""
      isDetails={false}
      itemList={Array []}
    />
  </div>
</div>
`;
